src/common/Path.h

```diff
@@ -0,0 +1,47 @@
+//
+// Copyright (c) 2021 The Aquarium Project Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+//
+
+#ifndef PATH_H
+#define PATH_H
+
+#include <string>
+#include <vector>
+
+class Path {
+public:
+  static Path getVoidPath();
+  static Path getExecutablePath();
+
+  explicit Path(const std::string &path);
+  template <typename T>
+  explicit Path(const T &path) : Path(std::string(path)) {}
+  ~Path();
+  bool isAbsolute() const;
+  bool isRelative() const;
+  Path &push(const Path &path);
+  template <typename T>
+  Path &push(const T &path) {
+    return push(Path(path));
+  }
+  Path &pop();
```

Comment:
If `push()` and `pop()` are too cryptic, is there a better candidate pair? `enter()`/`exit()`?

Reply:
enter() and exit() are also very weird for me. The problem for pop() is we usually get the popped value instead of the remaining part from calling it. 
How about append()/attach() and remove()/detach()/drop()?

Reply:
The advantage of `enter()` is that it sounds less misleading when the argument is an absolute path or a relative path prefixed with a different root name. I have mixed feeling on `append()`. It's better than others considering C++17 uses that name, and also worse due to the lack of an obvious antonym. Perhaps `drop()` or `strip()` is acceptable if we have to choose `append()`, but be aware it shares the same pitfall as `pop()`: the result for `"../.."` should be `"../../.."` instead of `".."`.

The biggest motivation of the path utility class is to serve the migration to Abseil, whereas a completely different proposal is to revert #124, as if we invented a "better" option parser than Abseil. Then nothing in 3rd-party needs further elimination, and at the same time nothing is blocked by this PR, as long as I can leave the legacy code as TBD until the arrival of C++17.

Reply:
I just recalled why `enter()` and `exit()` were turned down at the very beginning of this work: while it sounds good for "enter/exit `DIRECTORY`", it's improper to say "enter/exit `FILE`". So there's still no plausible names :(

_"There are only two hard things in computer science: cache invalidation and naming things."_